pkg/coscheduling/coscheduling.go

```diff
@@ -95,295 +77,208 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
+			waitTime = wait
+		}
+		if err == util.ErrorWaiting {
+			klog.Infof("Pod: %v is waiting to be scheduled to node: %v", core.GetNamedspacedName(pod), nodeName)
+			return framework.NewStatus(framework.Wait, ""), waitTime
+		}
+		// For pod not belongs to any groups
+		if err == util.ErrorNotMatched {
+			return framework.NewStatus(framework.Success, ""), 0
+		}
+		klog.Infof("cs.pgMgr.Permit error %v", err)
+		return framework.NewStatus(framework.Unschedulable, err.Error()), waitTime
 	}
 
-	klog.V(3).Infof("The count of PodGroup %v/%v/%v is up to minAvailable(%d) in Permit: current(%d)",
-		pod.Namespace, podGroupName, pod.Name, minAvailable, current)
-	cs.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
-		if waitingPod.GetPod().Namespace == namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
-			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
-			waitingPod.Allow(cs.Name())
-		}
-	})
+	klog.V(5).Infof("Pod requires pgName %v", fullName)
+	if ready {
+		cs.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+			if util.GetPodGroupFullName(waitingPod.GetPod()) == fullName {
+				klog.V(3).Infof("Permit allows the pod: %v", core.GetNamedspacedName(waitingPod.GetPod()))
+				waitingPod.Allow(cs.Name())
+			}
+		})
+		klog.V(3).Infof("Permit allows the pod: %v", core.GetNamedspacedName(pod))
+		return framework.NewStatus(framework.Success, ""), 0
+	}
 
-	return framework.NewStatus(framework.Success, ""), 0
+	return framework.NewStatus(framework.Wait, ""), waitTime
 }
 
-func (cs *Coscheduling) Reserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) *framework.Status {
+func (cs *Coscheduling) Reserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
 	return nil
 }
 
 // Unreserve rejects all other Pods in the PodGroup when one of the pods in the group times out.
 func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) {
-	pgInfo, _ := cs.getOrCreatePodGroupInfo(pod, time.Now())
-	if len(pgInfo.key) == 0 {
+	pgName, pgStatus := cs.pgMgr.GetPodGroup(pod)
+	if pgStatus == nil {
 		return
 	}
-	podGroupName := pgInfo.name
-	cs.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
-		if waitingPod.GetPod().Namespace == pod.Namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
-			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
+	cs.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		if waitingPod.GetPod().Namespace == pod.Namespace && waitingPod.GetPod().Labels[util.PodGroupLabel] == pgStatus.PodGroup.Name {
+			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", pgName, waitingPod.GetPod().Name)
 			waitingPod.Reject(cs.Name())
 		}
 	})
```

Comment:
Do do not see the obvious difference between your suggestion and the current one

Reply:
Seems you comment on a older version

Reply:
I was commenting on the behavior of how we reject pod - when a pod's waiting timed out (or whatever error), should we (1) reject that pod only, or (2) reject the whole PodGroup's pods, similar as what @yuanchen8911 raised at https://github.com/kubernetes-sigs/scheduler-plugins/pull/57#issuecomment-707384266.

@yuanchen8911 @denkensk ^^

Reply:
get it. 

In our production environment, we would usually meet: 
A job need 1000 pods, then, 900 pass check and wait, then the No.901 pod timeout. If the pod continue scheduling, the No.1 or No.100 may also expired. So I would like reject all though this may cost more time for scheduling the group.

But, we may add some optimization in the  feature, e.g. if the max allow scheduling time is 100s, the top 900 pods only consumes 10s, then we can only reject the current one.  The failed scheduling happens when  the 900  pods consume 95s or more, we reject all.